### PR TITLE
CRM-20286, added card type field on contribution search form

### PR DESCRIPTION
**Pull request by pradpnayak**
* [CRM-20286: Add card type field on search form](https://issues.civicrm.org/jira/browse/CRM-20286)

### CODE REVIEW

**Comment by pradpnayak** [comment]
@eileenmcnaughton can you please review this?

**Comment by eileenmcnaughton** [comment]
@pradpnayak I think this is good. I agree that your restructuring around the batch table is an improvement, and I was still able to do a search based on batch.

I have done a PR with a suggestion about changing the way you add the field to the search form.

Other than that, can you squash the commits into a single commit? Let me know if you are unsure how to

File: tests/phpunit/CRM/Contribute/Form/Search.php [hunk 1]
```diff
@@ -167,4 +167,109 @@ public function testBatchFilter() {
...
   }
 
+  /**
+   *  CRM-20286: Test CRM_Contribute_Form_Search Card type filters
+   */
+  public function testCardTypeFilter() {
+    $this->quickCleanup($this->_tablesToTruncate);
+    $contactID1 = $this->individualCreate(array(), 1);
+    $contactID2 = $this->individualCreate(array(), 2);
+    $Contribution1 = civicrm_api3('Contribution', 'create', array(
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Contribute/Form/Search.php]: minor point, in the tests it's better to use $this->callAPISuccess - it has better handling when there is a bug

File: tests/phpunit/CRM/Contribute/Form/Search.php [hunk 2]
```diff
@@ -167,4 +167,109 @@ public function testBatchFilter() {
...
+        'expected_count' => 0,
+        'expected_contribution' => array(),
+        'expected_qill' => 'Card Type In Amex',
+      ),
+      // Case 4: Search for ONLY those contributions which have Card Type as Visa or MasterCard
+      array(
+        'form_value' => array('financial_trxn_card_type' => array(1, 2)),
+        'expected_count' => 2,
+        'expected_contribution' => array($Contribution1['id'], $Contribution3['id']),
+        'expected_qill' => 'Card Type In Visa, MasterCard',
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Contribute/Form/Search.php]: including checks on the qill is good

File: CRM/Contribute/BAO/Query.php [hunk 3]
```diff
@@ -438,7 +440,15 @@ public static function whereClauseSingle(&$values, &$query) {
...
-        $query->_tables['contribution_financial_trxn'] = $query->_whereTables['contribution_financial_trxn'] = 1;
+        $query->_tables['civicrm_financial_trxn'] = $query->_whereTables['civicrm_financial_trxn'] = 1;
+        return;
+
+      case 'financial_trxn_card_type':
+        $query->_where[$grouping][] = CRM_Contact_BAO_Query::buildClause('civicrm_financial_trxn.card_type', $op, $value);
+        $query->_tables['civicrm_financial_trxn'] = $query->_whereTables['civicrm_financial_trxn'] = 1;
+        $query->_tables['civicrm_contribution'] = $query->_whereTables['civicrm_contribution'] = 1;
+        list($op, $value) = CRM_Contact_BAO_Query::buildQillForFieldValue('CRM_Financial_DAO_FinancialTrxn', 'card_type', $value, $op);
+        $query->_qill[$grouping][] = ts('%1 %2 %3', array(1 => ts('Card Type'), 2 => $op, 3 => $value));
```
Comment (monishdeb) [inline CRM/Contribute/BAO/Query.php]: Minor change 
```
-        $query->_qill[$grouping][] = ts('%1 %2 %3', array(1 => ts('Card Type'), 2 => $op, 3 => $value));
+        $query->_qill[$grouping][] = ts('Card Type %1 %2', array(1 => $op, 2 => $value));
```

File: tests/phpunit/CRM/Contribute/Form/Search.php (hunk 1 above)
Comment (monishdeb) [inline tests/phpunit/CRM/Contribute/Form/Search.php]: Agree.

**Comment by monishdeb** [comment]
The patch is working as expected, tested on my local 
image

PR needs two minor changes, one suggested by @eileenmcnaughton on unit-test and other by me on ```ts```.

**Pull request merged**